internal/core/v1beta1/distribution_service.go

```diff
@@ -0,0 +1,150 @@
+	"github.com/kanopy-platform/cdnvalidator/pkg/aws/cloudfront"
+)
+
+type DistributionService struct {
+	config     *config.Config
+	configFile string
+	cloudfront *cloudfront.Client
+	awsRegion  string
+	awsKey     string
+	awsSecret  string
+	timeout    time.Duration
+}
+
+func New(opts ...Option) (*DistributionService, error) {
+	var err error
+
+	d := &DistributionService{
+		config:  config.New(),
+		timeout: 30 * time.Second,
+	}
+
+	for _, opt := range opts {
+		opt(d)
+	}
+
+	// set up config
+	if d.configFile != "" {
+		if err := d.config.Watch(d.configFile); err != nil {
+			return nil, err
+		}
+	}
+
+	// set up AWS Cloudfront client
+	cfOpts := []cloudfront.Option{
+		cloudfront.WithAwsRegion(d.awsRegion),
+		cloudfront.WithStaticCredentials(d.awsKey, d.awsSecret),
+		cloudfront.WithTimeout(d.timeout),
+	}
+
+	d.cloudfront, err = cloudfront.New(cfOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return d, nil
+}
+
+func (d *DistributionService) getDistribution(ctx context.Context, distributionName string) (*config.Distribution, error) {
+	claims := core.GetClaims(ctx)
+	if len(claims) == 0 {
+		return nil, errors.New("no claims present")
+	}
+
+	// check user is entitled to the distributionName
+	entitlements := d.config.DistributionsFromClaims(claims)
+	distribution := d.config.Distribution(distributionName)
+
+	if _, ok := entitlements[distributionName]; !ok {
+		if distribution != nil {
+			// distribution exists but user is not entitled to it
+			return nil, NewInvalidationError(InvalidationUnauthorizedErrorCode, fmt.Errorf("distribution unauthorized"), distributionName)
+		} else {
+			return nil, NewInvalidationError(ResourceNotFoundErrorCode, fmt.Errorf("distribution %s not found", distributionName), distributionName)
+		}
+	}
+
+	return distribution, nil
+}
+
+func (d *DistributionService) List(ctx context.Context) (map[VanityDistributionName]Distribution, error) {
+	claims := core.GetClaims(ctx)
+	if len(claims) == 0 {
+		return nil, errors.New("no claims present")
+	}
+
+	ret := make(map[VanityDistributionName]Distribution)
+
+	distributions := d.config.DistributionsFromClaims(claims)
+	for name := range distributions {
+		distribution := d.config.Distribution(name)
+		ret[VanityDistributionName(name)] = Distribution{distribution.ID, distribution.Prefix}
+	}
+
+	return ret, nil
+}
+
+func (d *DistributionService) CreateInvalidation(ctx context.Context, distributionName string, paths []string) (*InvalidationResponse, error) {
+	distribution, err := d.getDistribution(ctx, distributionName)
+	if err != nil {
+		return nil, err
+	}
+
+	// check paths are valid
+	invalidPaths := make([]string, 0)
+
+	for _, path := range paths {
+		if !strings.HasPrefix(path, distribution.Prefix) {
+			invalidPaths = append(invalidPaths, path)
+		}
+	}
+	if len(invalidPaths) > 0 {
+		return nil, NewInvalidationError(InvalidationUnauthorizedErrorCode, fmt.Errorf("paths unauthorized"), invalidPaths)
+	}
```

Comment:
This check won't be necessary as the idea is that config will be the canonical place to indicate what prefix is allowed based on distribution. When getting the prefix from config, it will be prepended to whatever path the user enters, therefore the user won't enter a full path but a relative path that can formally be anything it wants,

i.e docs team will have a distribution set for `/docs` so if they enter `/*` or `/new/*`, this service should transform those into `/docs/*` and `/docs/new/*` respectively. We are essentially given a whole namespace per prefix to each entitlement.

This doesn't solve the problem of what happens if a path in config is not valid for the distribution, or doesn't exists. We might want to have validations on that or simply let it "fail" by not executing any action on invalid prefixes.

Reply:
It's out of scope of this service to know if a path actually exists on the CDN. As long as the user is entitled to perform the invalidation, it will go through no problem.  Cloudfront will not error.

Reply:
Agreed that checking if the path exists is out of scope here, I'm mainly asking what is the behaviour if the prefix doesn't exists, because a mistake in config, and a call is made to cloudfront?

Reply:
Cloudfront will just process the request and look for anything in it's cache that matches the path glob.  If that is nothing, due to a mistake in configuration, then it's a no-op.  From this services perspective it's a successful invalidation request.

Reply:
all clear

Reply:
Thanks Ricardo. I'll make the change so that `distribution.Prefix` is prepended to the passed in `paths`, and remove this check.